### ISSUE TITLE
[Bug] DeepSeek-V4 has slight accuracy issue

### ISSUE THREAD

**Issue by SYChen123** [sgl-project/sglang#28414]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

DeepSeek-V4 will generate plausible but wrong answers. It sometimes misses tokens or generates tokens with wrong order. It has huge affect on tool call scenarios. For example, in the response of tool call, the tokens generated is like "cd /path/to/slang/benc_mark". "sglang" and "benchmark" are not correctly returned. The agent will be likely trapped into the situation: get the inaccurate cmd => terminal error raised => llm fix the cmd => execute again.

### Reproduction

Launch Server
```
SGLANG_OPT_FP8_WO_A_GEMM=False \
  SGLANG_OPT_SWA_SPLIT_LEAF_ON_INSERT=1 \
  SGLANG_OPT_USE_CUSTOM_ALL_REDUCE_V2=1 \
  sglang serve \
    --trust-remote-code \
    --model-path ./models/deepseek-ai/DeepSeek-V4-Pro \
    --served-model-name ucloud-DeepSeek-V4-Pro \
    --tp 8 \
    --moe-runner-backend flashinfer_mxfp4 \
    --speculative-algo EAGLE \
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 4 \
    --chunked-prefill-size 8192 \
    --swa-full-tokens-ratio 0.1 \
    --mem-fraction-static 0.9 \
    --tool-call-parser deepseekv4 \
    --reasoning-parser deepseek-v4 \
    --enable-cache-report \
    --enable-metrics \
    --host 0.0.0.0 \
    --port 8000  > txt.log 2>&1 &
```

The test script is pasted below. This script sends a tool call request repeatedly and tells the model to send back the exact same command without changes. Finally the script will compare the expected command and the command that the model returns. 

Example usage: `python test.py 3600 32`
It means 3600 requests and 32 concurrency. It should be observed soon that the output generated is not the same as input prompt and some parts of the command is lost.

```
import concurrent.futures
import json
import os
import sys
import time
import urllib.request

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1800
WORKERS = int(sys.argv[2]) if len(sys.argv) > 2 else 32
LABEL = sys.argv[3] if len(sys.argv) > 3 else "stress"
URL = sys.argv[4] if len(sys.argv) > 4 else os.getenv("SGL_STRESS_URL", "http://127.0.0.1:8000/v1/chat/completions")
MODEL = sys.argv[5] if len(sys.argv) > 5 else os.getenv("SGL_STRESS_MODEL", "ucloud-DeepSeek-V4-Pro")
paths = [
"/upfs/xiaoyuan/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/whatcanIsaymambaout/workspace/usr/local/bin/cuda13.0/lib/site-packages/csy-sglang-ngram-v2/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/models/deepseek4/spec/ngram_worker/eagle_worker_v2/eagle_info_v2/Users/chensiyuan/LLMInfer/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/flashinfer/flashattention3/code/attention_backend/moe/triton/fused_moe/scheduler/managers/scheduler//upfs/xiaoyuan/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/upfs/xiaoyuan/code/sglang/upfs/abing/sglang/upfs/abing/sglang/qwen3-next/upfs/abing/vllm/vllm/.github/workflows/upfs/abing/vllm/vllm/benchmarks/fused_kernels/layernorm_rms_benchmarks/upfs/xiaoyuan/datasets/lmms-lab/flickr30k/datasets/qwen-image-edit/upfs/xiaoyuan/weights/Qwen/Qwen2.5-VL-72B-Instruct-AWQ/upfs/xiaoyuan/pkgs/.vscode-server/upfs/xiaoyuan/code/Mooncake/docs/source/python-api-reference/upfs/xiaoyuan/code/ufs_b/wav2lip/models/upfs/xiaoyuan/code/opencompass/data/piqa/upfs/xiaoyuan/code/llm-compressor/examples/quantization_w4a16/",
]
tool = {
    "type": "function",
    "function": {
        "name": "exec",
        "description": "Execute a shell command",
        "parameters": {
            "type": "object",
            "properties": {"command": {"type": "string"}, "timeout": {"type": "integer"}},
            "required": ["command"],
        },
    },
}

def one(i):
    path = paths[i % len(paths)]
    expected = f"cd {path} && pwd"
    payload = {
        "model": MODEL,
        "messages": [
            {"role": "system", "content": "You are a coding agent. Use exec for shell commands."},
            {"role": "user", "content": f"Call exec with this exact command and no changes: {expected}"},
        ],
        "tools": [tool],
        "temperature": 0,
        "max_tokens": 4096,
        "extra_body": {"chat_template_kwargs": {"thinking": True}},
    }
    req = urllib.request.Request(URL, data=json.dumps(payload).encode(), headers={"Content-Type": "application/json"})
    # print(f"create a request")
    try:
        with urllib.request.urlopen(req, timeout=240) as r:
            obj = json.loads(r.read())
    except Exception as e:
        return i, expected, "REQUEST_ERROR " + repr(e)
    try:
        # print(f"response: {obj}")
        args = obj["choices"][0]["message"]["tool_calls"][0]["function"]["arguments"]
        if isinstance(args, str):
            args = json.loads(args)
        cmd = args.get("command")
    except Exception as e:
        cmd = "PARSE_ERROR " + repr(e) + " " + json.dumps(obj)[:300]
    return i, expected, cmd

bad = []
start = time.time()
print(f"{LABEL}_start N={N} workers={WORKERS} url={URL} model={MODEL}", flush=True)
with concurrent.futures.ThreadPoolExecutor(max_workers=WORKERS) as ex:
    futures = [ex.submit(one, i) for i in range(N)]
    for fut in concurrent.futures.as_completed(futures):
        i, exp, cmd = fut.result()
        if cmd != exp:
            bad.append((i, exp, cmd))
            print("BAD", i, "EXP", exp, "GOT", cmd, flush=True)
print(f"{LABEL}_total {N} bad {len(bad)} elapsed {time.time() - start:.1f}", flush=True)
```

### Environment

Python: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA B200
GPU 0,1,2,3,4,5,6,7 Compute Capability: 10.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 570.172.08
PyTorch: 2.11.0+cu130
sglang: 0.0.0
sglang-kernel: 0.4.3
flashinfer_python: 0.6.12
flashinfer_cubin: 0.6.12
flashinfer_jit_cache: 0.6.12+cu130
triton: 3.6.0
transformers: 5.8.1
torchao: 0.17.0+cu130
numpy: 2.3.5
aiohttp: 3.14.1
fastapi: 0.137.1
huggingface_hub: 1.13.0
interegular: 0.3.3
modelscope: 1.36.3
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.4
python-multipart: 0.0.27
pyzmq: 27.1.0
uvicorn: 0.49.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.98.1
litellm: Module Not Found
torchcodec: 0.11.1+cu130
NVIDIA Topology: 
	[4mGPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	NIC2	NIC3	NIC4	NIC5	NIC6	NIC7	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NV18	SYS	SYS	SYS	SYS	SYS	SYS	NODE	PHB	138-183	3		N/A
GPU1	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NV18	SYS	SYS	SYS	SYS	SYS	SYS	PHB	NODE	138-183	3		N/A
GPU2	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NV18	SYS	SYS	SYS	SYS	NODE	PHB	SYS	SYS	92-137	2		N/A
GPU3	NV18	NV18	NV18	 X 	NV18	NV18	NV18	NV18	SYS	SYS	SYS	SYS	PHB	NODE	SYS	SYS	92-137	2		N/A
GPU4	NV18	NV18	NV18	NV18	 X 	NV18	NV18	NV18	SYS	SYS	NODE	PHB	SYS	SYS	SYS	SYS	46-91	1		N/A
GPU5	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NV18	SYS	SYS	PHB	NODE	SYS	SYS	SYS	SYS	46-91	1		N/A
GPU6	NV18	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NODE	PHB	SYS	SYS	SYS	SYS	SYS	SYS	0-45	0		N/A
GPU7	NV18	NV18	NV18	NV18	NV18	NV18	NV18	 X 	PHB	NODE	SYS	SYS	SYS	SYS	SYS	SYS	0-45	0		N/A
NIC0	SYS	SYS	SYS	SYS	SYS	SYS	NODE	PHB	 X 	NODE	SYS	SYS	SYS	SYS	SYS	SYS				
NIC1	SYS	SYS	SYS	SYS	SYS	SYS	PHB	NODE	NODE	 X 	SYS	SYS	SYS	SYS	SYS	SYS				
NIC2	SYS	SYS	SYS	SYS	NODE	PHB	SYS	SYS	SYS	SYS	 X 	NODE	SYS	SYS	SYS	SYS				
NIC3	SYS	SYS	SYS	SYS	PHB	NODE	SYS	SYS	SYS	SYS	NODE	 X 	SYS	SYS	SYS	SYS				
NIC4	SYS	SYS	NODE	PHB	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	 X 	NODE	SYS	SYS				
NIC5	SYS	SYS	PHB	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	 X 	SYS	SYS				
NIC6	NODE	PHB	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	 X 	NODE				
NIC7	PHB	NODE	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	SYS	NODE	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_4
  NIC1: mlx5_5
  NIC2: mlx5_6
  NIC3: mlx5_7
  NIC4: mlx5_8
  NIC5: mlx5_9
  NIC6: mlx5_10
  NIC7: mlx5_11

ulimit soft: 1024

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Hi @SYChen123 

> paths = [
"/upfs/xiaoyuan/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/whatcanIsaymambaout/workspace/usr/local/bin/cuda13.0/lib/site-packages/csy-sglang-ngram-v2/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/sglang/python/srt/layers/attention/vision/triton/models/deepseek4/spec/ngram_worker/eagle_worker_v2/eagle_info_v2/Users/chensiyuan/LLMInfer/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/flashinfer/flashattention3/code/attention_backend/moe/triton/fused_moe/scheduler/managers/scheduler//upfs/xiaoyuan/code/sglang/python/sglang/srt/openai_api/deepseek_v4_tool_parser_exact_copy_regression_sglang_path_with_long_unique_suffix/upfs/xiaoyuan/code/sglang/upfs/abing/sglang/upfs/abing/sglang/qwen3-next/upfs/abing/vllm/vllm/.github/workflows/upfs/abing/vllm/vllm/benchmarks/fused_kernels/layernorm_rms_benchmarks/upfs/xiaoyuan/datasets/lmms-lab/flickr30k/datasets/qwen-image-edit/upfs/xiaoyuan/weights/Qwen/Qwen2.5-VL-72B-Instruct-AWQ/upfs/xiaoyuan/pkgs/.vscode-server/upfs/xiaoyuan/code/Mooncake/docs/source/python-api-reference/upfs/xiaoyuan/code/ufs_b/wav2lip/models/upfs/xiaoyuan/code/opencompass/data/piqa/upfs/xiaoyuan/code/llm-compressor/examples/quantization_w4a16/",
]

Paths should be separated by comma.

**Comment by SYChen123** [comment]
@yiakwy-xpu-ml-framework-team There's only 1 path in the test script

**Comment by zhangxiaolei123456** [comment]
@SYChen123 this PR will fix your issue, you can try https://github.com/sgl-project/sglang/pull/28612

**Comment by SYChen123** [comment]
> [@SYChen123](https://github.com/SYChen123) this PR will fix your issue, you can try [#28612](https://github.com/sgl-project/sglang/pull/28612)

Hi @zhangxiaolei123456 Thanks for your comments. I wonder if this PR will take effect on the default c128 (no online compression)? I didn't turn on the online c128 when launching server.

**Comment by zhangxiaolei123456** [comment]
This pr https://github.com/sgl-project/sglang/pull/26471 affect no online compression

**Comment by Fridge003** [comment]
@SYChen123 This PR also fixed an accuracy bug we observed before:
https://github.com/sgl-project/sglang/pull/26766
Maybe you can try rebasing to latest main and check whether this bug is still there

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Hi @Fridge003 I believe the root answer is predicting draft tokens in massive parallel will corrupt kV cache (tree mask kv cache).

Morover, I am confident that exsiting MTP method is not fundamentally not LossLess in theory for 3 reasons and we will have a work on it:

1.  Consider sequence modeling problem p_target(T3|T1, T2), p_draft(T3'|T1, T2). Traditoinal refusion sampling working on comparing p1=p_target(T3|T1, T2), and p3 = p_draft(T3|T1, T2), with an uniform random variable x (0<x<1), i.e.  x < p1/(c*p3) (p1 < (c*p3))

1.1 However exsiting methods   compares p1 and p4:

> p4 = p_draft(T3'|T1, T2)

Since by no means we know that it is T3,  instead of T3' the token we wanted, neither can we recall the true posterior distribution (the logits we computed from target model) from proposal posterial distribution.


2. tradition rejection sampling is not built for sequence modeling. 

 sine p_draft is not the one derived from p_target, if we accept as long as x is not in refusion region, we immedately change the probability of target model:

> p = p_target(T4 | T1, T2, T3) != p_target(T4 | T1, T2, T3')


3. Tree-Mask KV Cache Pollution when predicting draft tokens in parallel

Like this exmaple.

cc @SYChen123
